File: Homework/Objective-C/HW30/HW30/ViewController.m
//
//  ViewController.m
//  HW30
//
//  Created by Руслан on 05.11.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

#import "ViewController.h"
#import "RTRandomColor.h"
#import "RTStudent.h"

@interface ViewController ()

@property (strong, nonatomic) NSMutableArray *randomColors;
@property (strong, nonatomic) NSMutableArray *studentsSortedByNames;
@property (strong, nonatomic) NSMutableArray *studentsSortedByMarks;

@property (strong, nonatomic) NSMutableArray *studentsLowMarks;
@property (strong, nonatomic) NSMutableArray *studentsMediocreMarks;
@property (strong, nonatomic) NSMutableArray *studentsGoodMarks;
@property (strong, nonatomic) NSMutableArray *studentsExcellentMarks;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    
    self.randomColors = [[NSMutableArray alloc] init];
    
    for (int i = 0; i < 5; i++) {
        RTRandomColor *randomColor = [[RTRandomColor alloc] initWithColorAndName];
        [self.randomColors addObject:randomColor];
    }
    
    
    //self.studentsSortedByNames = [[NSMutableArray alloc] init];
    self.studentsSortedByMarks = [[NSMutableArray alloc] init];
    
    self.studentsLowMarks = [[NSMutableArray alloc] init];
    self.studentsMediocreMarks = [[NSMutableArray alloc] init];
    self.studentsGoodMarks = [[NSMutableArray alloc] init];
    self.studentsExcellentMarks = [[NSMutableArray alloc] init];
    
    for (int i = 0; i < 25; i++) {
        RTStudent *student = [[RTStudent alloc] initWithFullNameAndAvMark];
        //[self.studentsSortedByNames addObject:student];
        [self.studentsSortedByMarks addObject:student];
        
        if (student.averageMark < 2.5) {
            [self.studentsLowMarks addObject:student];
        } else if (student.averageMark < 3.5) {
            [self.studentsMediocreMarks addObject:student];
        } else if (student.averageMark < 4.5) {
            [self.studentsGoodMarks addObject:student];
        } else {
            [self.studentsExcellentMarks addObject:student];
        }
    }
    
    //self.studentsSortedByNames = [self sortArrayByNames:self.studentsSortedByNames];
    
    self.studentsLowMarks = [self sortArrayByNames:self.studentsLowMarks];
    self.studentsMediocreMarks = [self sortArrayByNames:self.studentsMediocreMarks];
    self.studentsGoodMarks = [self sortArrayByNames:self.studentsGoodMarks];
    self.studentsExcellentMarks = [self sortArrayByNames:self.studentsExcellentMarks];
    
    [self.studentsSortedByMarks sortUsingComparator:^NSComparisonResult(RTStudent *obj1, RTStudent *obj2) {
        
        if (obj1.averageMark >= obj2.averageMark) {
            return NSOrderedAscending;
        } else {
            return NSOrderedDescending;
        }
        
    }];
    
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Helpful functions

- (NSMutableArray *) sortArrayByNames:(NSMutableArray *)array {
    
    [array sortUsingComparator:^NSComparisonResult(RTStudent *obj1, RTStudent *obj2) {
        
        if ([obj1.lastName isEqualToString:obj2.lastName]) {
            return [obj1.firstName compare:obj2.firstName];
        } else {
            return [obj1.lastName compare:obj2.lastName];
        }
    }];
    
    return array;
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 5;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    switch (section) {
        case 0:
            return [self.studentsLowMarks count];
            break;
        case 1:
            return [self.studentsMediocreMarks count];
            break;
        case 2:
            return [self.studentsGoodMarks count];
            break;
        case 3:
            return [self.studentsExcellentMarks count];
            break;
        case 4:
            return [self.randomColors count];
            break;
        default:
            return 0;
            break;
    }
}

- (nullable NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    
    switch (section) {
        case 0:
            return @"Low mark students";
            break;
        case 1:
            return @"Mediocre students";
            break;
        case 2:
            return @"Good mark students";
            break;
        case 3:
            return @"Excellent mark students";
            break;
        case 4:
            return @"Colors";
            break;
        default:
            return @"Default";
            break;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *student = @"student";
    static NSString *color = @"color";
    
    UITableViewCell *cell = [[UITableViewCell alloc] init];

    if (indexPath.section == 4) {
        
        cell = [tableView dequeueReusableCellWithIdentifier:color];
        
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:color];
        }
        
        cell.backgroundColor = ((RTRandomColor *)[self.randomColors objectAtIndex:indexPath.row]).color;
        cell.textLabel.textColor = [UIColor whiteColor];
        cell.textLabel.shadowColor = [UIColor blackColor];
        cell.textLabel.textAlignment = NSTextAlignmentCenter;
        cell.textLabel.text = ((RTRandomColor *)[self.randomColors objectAtIndex:indexPath.row]).name;
        
    } else {
        
        cell = [tableView dequeueReusableCellWithIdentifier:student];
        
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:student];
        }
        
        RTStudent *student = [[RTStudent alloc] init];
        
        switch (indexPath.section) {
            case 0:
                student = [self.studentsLowMarks objectAtIndex:indexPath.row];
                cell.textLabel.textColor = [UIColor redColor];
                break;
            case 1:
                student = [self.studentsMediocreMarks objectAtIndex:indexPath.row];
                cell.textLabel.textColor = [UIColor orangeColor];
                break;
            case 2:
                student = [self.studentsGoodMarks objectAtIndex:indexPath.row];
                cell.textLabel.textColor = [UIColor magentaColor];
                break;
            case 3:
                student = [self.studentsExcellentMarks objectAtIndex:indexPath.row];
                cell.textLabel.textColor = [UIColor greenColor];
                break;
            default:
                break;
        }
        
        
        cell.textLabel.text = student.fullName;
        cell.detailTextLabel.text = [NSString stringWithFormat:@"%1.1f", student.averageMark];
    }
    
    /*
    CGFloat r = (CGFloat)(arc4random() % 256) / 255.f;
    CGFloat g = (CGFloat)(arc4random() % 256) / 255.f;
    CGFloat b = (CGFloat)(arc4random() % 256) / 255.f;
    
    cell.backgroundColor = [UIColor colorWithRed:r green:g blue:b alpha:1.f];
    cell.textLabel.textColor = [UIColor whiteColor];
    cell.textLabel.shadowColor = [UIColor blackColor];
    cell.textLabel.text = [NSString stringWithFormat:@"%ld. RGB(%d, %d, %d)",
                           indexPath.row, (int)(r * 255), (int)(g * 255), (int)(b * 255)];
    cell.textLabel.textAlignment = NSTextAlignmentCenter;
    */
    /*
    cell.backgroundColor = ((RTRandomColor *)[self.randomColors objectAtIndex:indexPath.row]).color;
    cell.textLabel.textColor = [UIColor whiteColor];
    cell.textLabel.shadowColor = [UIColor blackColor];
    cell.textLabel.textAlignment = NSTextAlignmentCenter;
    cell.textLabel.text = ((RTRandomColor *)[self.randomColors objectAtIndex:indexPath.row]).name;
    */

    return cell;
}


@end

File: Homework/Swift/HW30/HW30/ViewController.swift
//
//  ViewController.swift
//  HW30
//
//  Created by Руслан on 06.11.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDataSource {
    
// MARK: - Private properties
    
    private var randomColors = [RandomColor]()
    private var allStudents  = [Student]()
    
    private var studentsLowMarks        = [Student]()
    private var studentsMediocreMarks   = [Student]()
    private var studentsGoodMarks       = [Student]()
    private var studentsExcellentMarks  = [Student]()
    
// MARK: - TableView and viewDidLoad

    @IBOutlet weak var tableView: UITableView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // colors
        for _ in 0..<5 {
            let color = RandomColor()
            randomColors.append(color)
        }
        // students
        for _ in 0..<25 {
            
            let student = Student()
            allStudents.append(student)
            
            if student.averageMark < 2.5 {
                studentsLowMarks.append(student)
            } else if student.averageMark < 3.5 {
                studentsMediocreMarks.append(student)
            } else if student.averageMark < 4.5 {
                studentsGoodMarks.append(student)
            } else {
                studentsExcellentMarks.append(student)
            }
        }
        
        sortedByNames(&studentsLowMarks)
        sortedByNames(&studentsMediocreMarks)
        sortedByNames(&studentsGoodMarks)
        sortedByNames(&studentsExcellentMarks)
        
        allStudents = allStudents.sorted() { student1, student2 in
            return student1.averageMark > student2.averageMark
        }
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
// MARK: - Helpful functions
    
    func sortedByNames(_ students: inout [Student]) {
        
        students = students.sorted() { student1, student2 in
            
            if student1.lastName == student2.lastName {
                return student1.firstName > student2.firstName
            } else {
                return student1.lastName > student2.lastName
            }
        }
    }
    
// MARK: - UITableViewDataSource
    
    public func numberOfSections(in tableView: UITableView) -> Int {
        return 5
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        switch section {
        case 0:
            return studentsLowMarks.count
        case 1:
            return studentsMediocreMarks.count
        case 2:
            return studentsGoodMarks.count
        case 3:
            return studentsExcellentMarks.count
        case 4:
            return randomColors.count
        default:
            return 1
        }
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        
        switch section {
        case 0:
            return "Low mark students"
        case 1:
            return "Mediocre students"
        case 2:
            return "Good mark students"
        case 3:
            return "Excellent mark students"
        case 4:
            return "Colors"
        default:
            return "Default"
        }
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let color = "color"
        let student = "student"
        
        var cell: UITableViewCell?
        
        if indexPath.section == 4 {
            
            cell = tableView.dequeueReusableCell(withIdentifier: color)
            
            if cell == nil {
                cell = UITableViewCell.init(style: .default, reuseIdentifier: color)
            }
            
            cell?.backgroundColor = randomColors[indexPath.row].color
            cell?.textLabel?.textColor = UIColor.white
            cell?.textLabel?.shadowColor = UIColor.black
            cell?.textLabel?.textAlignment = .center
            cell?.textLabel?.text = randomColors[indexPath.row].name
            
            return cell!
            
        } else {
            
            cell = tableView.dequeueReusableCell(withIdentifier: student)
            
            if cell == nil {
                cell = UITableViewCell.init(style: .value1, reuseIdentifier: student)
            }
            
            let student: Student
            
            switch indexPath.section {
            case 0:
                student = studentsLowMarks[indexPath.row]
                cell?.textLabel?.textColor = UIColor.red
            case 1:
                student = studentsMediocreMarks[indexPath.row]
                cell?.textLabel?.textColor = UIColor.orange
            case 2:
                student = studentsGoodMarks[indexPath.row]
                cell?.textLabel?.textColor = UIColor.magenta
            default:
                student = studentsExcellentMarks[indexPath.row]
                cell?.textLabel?.textColor = UIColor.green
            }
            
            cell?.textLabel?.text = student.fullName
            cell?.detailTextLabel?.text = "\(student.averageMark)"
            
            return cell!
        }
    }
}
